src/sparsezoo/analysis/Analysis.py

```diff
@@ -0,0 +1,221 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+from typing import Dict, List, Optional
+
+import numpy as np
+import onnx
+from onnx import ModelProto
+
+from pydantic import BaseModel, Field
+from sparsezoo.analysis.helpers import (
+    get_layer_and_op_counts,
+    get_layer_param,
+    get_node_four_block_sparsity,
+    get_node_num_four_block_zeros_and_size,
+    get_node_num_zeros_and_size,
+    get_node_sparsity,
+    get_zero_point,
+    is_parameterized_prunable_layer,
+    is_quantized_layer,
+    is_sparse_layer,
+)
+
+
+__all__ = [
+    "NodeAnalysis",
+    "ModelAnalysis",
+]
+
+
+def map_and_accumulate(array, map_fn, acc_fn):
+    """
+    Helper function to apply a map function to each element in the array,
+    then apply the accumulate function to the whole array
+
+    :param: array to apply map function to
+    :param: map function to apply to the array
+    :param: accumulator function to apply after mapping
+
+    :return: The result of applying map function and then accumulator function
+    """
+    return acc_fn(list(map(map_fn, array)))
+
+
+class NodeAnalysis(BaseModel):
+    """
+    Pydantic model for the analysis of a node within a model
+    """
+
+    name: str = Field(description="This node's name")
+    sparsity: float = Field(description="Proportion of zeros in this node's parameter")
+    four_block_sparsity: float = Field(
+        description="Proportion of zero blocks in this node's parameter"
+    )
+    param_size: int = Field(description="Size of this node's parameter")
+    num_sparse_values: int = Field(
+        description="Number of zeros in this node's parameter"
+    )
+    num_four_blocks: int = Field(
+        description="Number of four blocks in this node's parameter"
+    )
+    num_sparse_four_blocks: int = Field(
+        description="Number of four blocks with all zeros in this node's parameter"
+    )
+    quantized_layer: bool = Field(description="Does this node have quantized weights")
+    zero_point: int = Field(
+        description="Node zero point for quantization, default zero"
+    )
+    dtype: Optional[str] = Field(
+        description="This node's parameter's data type", default=None
+    )
+
+
+class ModelAnalysis(BaseModel):
+    """
+    Pydantic model for analysis of a model
+    """
+
+    layer_counts: Dict[str, int] = Field(description="Overview of layers", default={})
+    op_counts: Dict[str, int] = Field(
+        description="Overview of operations (nodes that are not layers)", default={}
+    )
+    num_nodes: int = Field(description="Total number of nodes")
+    num_layers: int = Field(description="Total number of layers")
+    num_operations: int = Field(
+        description="Total number of operations (nodes that are not layers)"
+    )
```

Comment:
we can leave these out the dicts provide better information than this with about the same effort of looking through